### PR TITLE
ENH: Cleaner and faster trip generation

### LINKED ISSUE #219: Implement a cleaner and faster version of trip generation
Current code for trip generation is complicated. It would be great to have a cleaner appraoch that can be vectorized.

### PR DESCRIPTION
**Pull request by henrymartin1**
fixes #219

Main idea of the new implementation is that we assign a trip_ids to a combined dataframe of triplegs and staypoints so that we are able to create trips using a groupby statement that is based on the assigned trip id.

The implementation itself has several main steps:

0. Merge triplegs and staypoints into a single dataframe and sort them by [user_id, started_at]
1. identify start of new trips (activities, user change, temporal gaps) 
2. assign trip ids to all triplegs/staypoints that belong to a trip
3. delete activities as they are not part of trips
4. group by trip_id to create trips
5. add activities again to trip dataframe to assign origin_staypoint_id and destination_staypoint_id
6. delete activities again
7. Do some id stuff

Some tests are already passing but I have added the remaining todos in the code

### CODE REVIEW

**Comment by bifbof** [comment]
There is one thing I am not sure about, if we filter out trips without triplegs as we say in the docstring.

Maybe in another pull request we should write a test for that.

File: trackintel/preprocessing/triplegs.py [hunk 1]
```diff
@@ -113,30 +115,32 @@ def generate_trips(spts, tpls, gap_threshold=15):
...
     # start new trip if there is a new activity
     condition_new_activity = spts_tpls["last_activity"]
 
     # gap conditions
+    # start new trip after a gap
     gap = (spts_tpls["started_at_next"] - spts_tpls["finished_at"]) > gap_threshold
     condition_time_gap = gap.shift(1, fill_value=False)  # trip starts on next entry
-    # no gaps between activities
+
+    # no gaps between activities #todo: please elaborate condition further
```
Comment (bifbof) [inline trackintel/preprocessing/triplegs.py:137]: ^

File: trackintel/preprocessing/triplegs.py [hunk 2]
```diff
@@ -145,31 +149,40 @@ def generate_trips(spts, tpls, gap_threshold=15):
     spts_tpls_no_act = spts_tpls[~spts_tpls["activity"]]
     spts_tpls_only_act = spts_tpls[spts_tpls["activity"]]
 
+    # todo: maybe move this part down to where you need it
```
Comment (bifbof) [inline trackintel/preprocessing/triplegs.py:152]: ^

File: trackintel/preprocessing/triplegs.py [hunk 3]
```diff
@@ -145,31 +149,40 @@ def generate_trips(spts, tpls, gap_threshold=15):
...
     # add gaps as activities, to simplify id assignment later.
     gaps = pd.DataFrame(spts_tpls.loc[gap, "user_id"])
     gaps["started_at"] = spts_tpls.loc[gap, "finished_at"] + gap_threshold / 2
     gaps[["type", "activity", "new_trip"]] = ["gap", True, True]
 
     # same for user changes
     user_change = pd.DataFrame(spts_tpls.loc[condition_new_user, "user_id"])
+    # todo: if the started_at will be used for sorting then the time difference should be as small as possible
```
Comment (bifbof) [inline trackintel/preprocessing/triplegs.py:160]: It should not make any difference as we first sort by `user_id`.

File: trackintel/preprocessing/triplegs.py [hunk 4]
```diff
@@ -145,31 +149,40 @@ def generate_trips(spts, tpls, gap_threshold=15):
...
     user_change["started_at"] = spts_tpls.loc[condition_new_user, "started_at"] - gap_threshold / 2
     user_change[["type", "activity", "new_trip"]] = ["user_change", True, True]
 
     trips_grouper = spts_tpls_no_act.groupby("trip_id")
+    # todo: for me using `np.array` raised an exception: "ValueError: Must produce aggregated value"
```
Comment (bifbof) [inline trackintel/preprocessing/triplegs.py:165]: That is strange. Maybe they updated that in a newer pandas version. https://stackoverflow.com/questions/16975318/pandas-aggregate-when-column-contains-numpy-arrays

File: trackintel/preprocessing/triplegs.py [hunk 5]
```diff
@@ -221,7 +234,7 @@ def _seperate_ids(row):
     # assign trip_id to tpls
     temp = trips.explode("tpls")
     temp.index = temp["tpls"]
-    temp = temp[temp["spts"].notna()]
+    temp = temp[temp["spts"].notna()] #todo: what is this dropping exactly? Should be tpls?
```
Comment (bifbof) [inline trackintel/preprocessing/triplegs.py:237]: Yes, a mix-up. Although I am suprised that it didn't throw an error in the tests.

File: trackintel/preprocessing/triplegs.py [hunk 6]
```diff
@@ -86,11 +86,13 @@ def generate_trips(spts, tpls, gap_threshold=15):
     # If the triplegs already have a column "trip_id", we drop it
     if "trip_id" in tpls:
         tpls.drop(columns="trip_id", inplace=True)
+        #todo: should we add a warning here?
```
Comment (bifbof) [inline trackintel/preprocessing/triplegs.py:89]: If we would alter it in place I would be totally for it. But when returning a copy is that really necessary?

File: trackintel/preprocessing/triplegs.py [hunk 7]
```diff
@@ -145,31 +149,40 @@ def generate_trips(spts, tpls, gap_threshold=15):
...
+        row_id = np.array(row["id"])
+        t = row_type == "tripleg"
+        tpls_ids = row_id[t]
+        spts_ids = row_id[np.logical_not(t)]
         return [spts_ids, tpls_ids]
 
     trips[["spts", "tpls"]] = trips.apply(_seperate_ids, axis=1, result_type="expand")
 
     # save trip_id as column
+    # todo: I don't understand this. The trip id is the index. Therefore the reset operation creates the column "trip
```
Comment (bifbof) [inline trackintel/preprocessing/triplegs.py:182]: The groupby deletes empty trips, so we may have fewer trips than originally planned. The reason for this is not quite clear to me so far. So this reset_index is here to make sure that our trip index really goes from 0 to n-1. I'll have a look into it.

**Review by bifbof — CHANGES_REQUESTED** [review]
Thanks Henry for the review. As the Henry is the author of the PR he cannot review it. So I do that to resolve and discuss the review points more easily.

**Review by henrymartin1 — COMMENTED** [review]
Yes, thanks for changing the status. This "not being able to review it" is actually a bit annoying ;-)
I can actually add comments on github and planned to add them additionally to my in code review (as you did). But you were too fast ;-)
I added commented on your comments and copied the explicit todo message.

File: trackintel/preprocessing/triplegs.py (hunk 1 above)
Comment (bifbof) [reply trackintel/preprocessing/triplegs.py:137]: It is an artifact from the time there where still intermediate activities. I think these aren't necessary anymore.

File: trackintel/preprocessing/triplegs.py (hunk 3 above)
Comment (bifbof) [reply trackintel/preprocessing/triplegs.py:160]: A new data "gap" is always before the first data of this specific user and therefor behind everything of the last user. The order is never corrupted as the user id stays the same. For example
| user id | timestamp |
| ------- | --------- |
| 0       | 4000      |
| 1       | 1         |

Lets add two new user "gaps".
| user id | timestamp |
| ------- | --------- |
| 0       | 4000 - 1  |
| 0       | 4000      |
| 1       | 1 - 1     |
| 1       | 1         |

Here we used 1 one as subtrahend but the value doesn't matter.

File: trackintel/preprocessing/triplegs.py (hunk 4 above)
Comment (bifbof) [reply trackintel/preprocessing/triplegs.py:165]: Good fix. :)

File: trackintel/preprocessing/triplegs.py (hunk 7 above)
Comment (bifbof) [reply trackintel/preprocessing/triplegs.py:182]: So for example I've we have `activity - gap - tripleg` we have an empty trip as the next trip already starts on the tripleg so the activity has no "non" activities we can aggregate &rarr; therefor we need to reindex to get the trip id without the empty trips.

**Comment by henrymartin1** [comment]
Is this now ready for review again? It is a bit messy with not being able to be requested as a reviewer for this PR ;-)

**Comment by bifbof** [comment]
> Is this now ready for review again? It is a bit messy with not being able to be requested as a reviewer for this PR ;-)

Yes it should be good again. :)

File: trackintel/preprocessing/triplegs.py (hunk 3 above)
Comment (henrymartin1) [reply trackintel/preprocessing/triplegs.py]: Agree. 👍

File: trackintel/preprocessing/triplegs.py (hunk 6 above)
Comment (henrymartin1) [reply trackintel/preprocessing/triplegs.py]: 👍

File: trackintel/preprocessing/triplegs.py (hunk 7 above)
Comment (henrymartin1) [reply trackintel/preprocessing/triplegs.py]: 👍

**Comment by henrymartin1** [comment]
@hong2223 Ready for your review!

**Comment by bifbof** [comment]
Thanks Henry for the cleanup! :)

**Review by bifbof — APPROVED** [review]

File: trackintel/preprocessing/triplegs.py [hunk 8]
```diff
@@ -89,283 +86,204 @@ def generate_trips(stps_input, tpls_input, gap_threshold=15, print_progress=Fals
...
+    spts_tpls["id"] = spts_tpls.index
+
+    spts_tpls.sort_values(by=["user_id", "started_at"], inplace=True)
+    spts_tpls.reset_index(inplace=True, drop=True)
+
+    # delete intermediate activities
+    # intermediate activities are activities that are surrounded by other activities
+    # therefore they are neither at the end, at the start, nor within a trip
+    first_activities, inter_activities, last_activities = _get_activity_masks(spts_tpls)
+    spts_tpls["first_activity"] = first_activities
```
Comment (hong2223) [inline trackintel/preprocessing/triplegs.py:131]: why do we need this? `spts_tpls["first_activity"]` never used and eventually deleted (in Line 213)

File: trackintel/preprocessing/triplegs.py [hunk 9]
```diff
@@ -89,283 +86,204 @@ def generate_trips(stps_input, tpls_input, gap_threshold=15, print_progress=Fals
...
+    # conditions for new trip
+    # start new trip if the user changes
+    condition_new_user = spts_tpls["user_id"] != spts_tpls["user_id"].shift(1)
+
+    # start new trip if there is a new activity
+    condition_new_activity = spts_tpls["last_activity"]
+
+    # gap conditions
+    # start new trip after a gap
+    gap = (spts_tpls["started_at_next"] - spts_tpls["finished_at"]) > gap_threshold
```
Comment (hong2223) [inline trackintel/preprocessing/triplegs.py:145]: Current `spts_tpls` is after removing the `inter_activity` (line 133). Isn't that the spts are now filtered, and spt information might be lost?

File: trackintel/preprocessing/triplegs.py [hunk 10]
```diff
@@ -89,283 +86,204 @@ def generate_trips(stps_input, tpls_input, gap_threshold=15, print_progress=Fals
...
+    spts_tpls_no_act = spts_tpls[~spts_tpls["activity"]]
+    spts_tpls_only_act = spts_tpls[spts_tpls["activity"]]
+
+    trips_grouper = spts_tpls_no_act.groupby("trip_id")
+    trips = trips_grouper.agg({"user_id": "mean", "started_at": min, "finished_at": max, "type": list, "id": list})
+
+    def _seperate_ids(row):
+        """Split aggregated ids into staypoints and triplegs arrays."""
+        row_type = np.array(row["type"])
+        row_id = np.array(row["id"])
```
Comment (hong2223) [inline trackintel/preprocessing/triplegs.py:166]:
```suggestion
        row_type = row["type"].values
        row_id = row["id"].values
```

File: trackintel/preprocessing/triplegs.py [hunk 11]
```diff
@@ -89,283 +86,204 @@ def generate_trips(stps_input, tpls_input, gap_threshold=15, print_progress=Fals
...
+        t = row_type == "tripleg"
+        tpls_ids = row_id[t]
+        spts_ids = row_id[np.logical_not(t)]
```
Comment (hong2223) [inline trackintel/preprocessing/triplegs.py:169]:
```suggestion
        spts_ids = row_id[~t]
```

File: trackintel/preprocessing/triplegs.py [hunk 12]
```diff
@@ -89,283 +86,204 @@ def generate_trips(stps_input, tpls_input, gap_threshold=15, print_progress=Fals
...
+        return [spts_ids, tpls_ids]
+
+    trips[["spts", "tpls"]] = trips.apply(_seperate_ids, axis=1, result_type="expand")
+
+    # recount trips ignoring empty trips and save trip_id as for id assignment.
+    trips.reset_index(inplace=True)
```
Comment (hong2223) [inline trackintel/preprocessing/triplegs.py:175]: the original index is not dropped? i.e., `trips.reset_index(inplace=True, drop=True)`

File: trackintel/preprocessing/triplegs.py [hunk 13]
```diff
@@ -89,283 +86,204 @@ def generate_trips(stps_input, tpls_input, gap_threshold=15, print_progress=Fals
...
+
+    # same for user changes
+    user_change = pd.DataFrame(spts_tpls.loc[condition_new_user, "user_id"])
+    user_change["started_at"] = spts_tpls.loc[condition_new_user, "started_at"] - gap_threshold / 2
+    user_change[["type", "activity", "new_trip"]] = ["user_change", True, True]
+
+    # merge with activities
+    trips.drop(columns=["type", "id"], inplace=True)  # make space so no overlap with activity "id"
+    spts_tpls_only_act.drop(columns=["trip_id"])  # no overlap of activity trip_ids to real trip_ids in trips
+    trips = pd.concat((trips, spts_tpls_only_act, gaps, user_change), axis=0, ignore_index=True)
```
Comment (hong2223) [inline trackintel/preprocessing/triplegs.py:191]: I am confused by the `type`, `id` and `trip_id` columns for each of the three different df at this point. Is there a way to increase the readability here?

Also, the `trips` here is not yet really trips? In line 204 we get the staypoints from "trips". I guess here we are working on activitySpts_trips?

File: tests/preprocessing/test_triplegs.py [hunk 14]
```diff
@@ -146,6 +147,26 @@ def test_generate_trips_dtype_consistent(self):
         assert stps["next_trip_id"].dtype == "Int64"
         assert tpls["trip_id"].dtype == "Int64"
 
+    def test_compare_to_old_trip_function(self):
```
Comment (hong2223) [inline tests/preprocessing/test_triplegs.py:150]: would be nice to run this function on Geolife. I can do that after the other comments are answered

**Review by hong2223 — CHANGES_REQUESTED** [review]

File: trackintel/preprocessing/triplegs.py (hunk 9 above)
Comment (bifbof) [reply trackintel/preprocessing/triplegs.py:145]: No It should not, as no trip can start or end on these intermediate activities.
But I think I do not fully understand your question.

File: trackintel/preprocessing/triplegs.py (hunk 10 above)
Comment (bifbof) [reply trackintel/preprocessing/triplegs.py:166]: `row["type"]` is a list and list don't have an attribute `values`. But I agree that casting to numpy array is not that clean.

File: trackintel/preprocessing/triplegs.py (hunk 11 above)
Comment (bifbof) [reply trackintel/preprocessing/triplegs.py:169]: The `np.logical_not(t)` is here for singletons aka `True`. If we would use the bitwise operator `~` for those we get the value -2 (And -1 for False). I guess that is because of some type-casting in python.
We get singletons if there is only one entry in the list.

File: trackintel/preprocessing/triplegs.py (hunk 13 above)
Comment (bifbof) [reply trackintel/preprocessing/triplegs.py:191]: I renamed `"id"` to `"spts_tpls_id"` and the temporary trip_ids to `"temp_trip_id"` and the final trip_id remains `"trip_id"`. I left `"type"` as the name because I reuse that for gaps, new_user activities and find the name appropriate and not so confusing when you change the others.

Yes, these are really trips with activities, I renamed it accordingly to show the differences.
I think with these changes the function will be much cleaner. Thanks!

File: trackintel/preprocessing/triplegs.py (hunk 11 above)
Comment (bifbof) [reply trackintel/preprocessing/triplegs.py]: I doubled checked you were right, that was an legacy issue from aggregating directly to numpy arrays.

File: trackintel/preprocessing/triplegs.py (hunk 9 above)
Comment (bifbof) [reply trackintel/preprocessing/triplegs.py]: I cleaned up the code a bit there now it isn't necessary anymore to filter for `inter_activities`.

**Comment by bifbof** [comment]
The linting check keeps failing and my local black don't want to change anything. :)

**Review by hong2223 — CHANGES_REQUESTED** [review]
Thanks for the changes after the last review. 

I compared the trip generation using the old and new function for Geolife user 1, the trip_id 552 trip from `2009-06-27 10:55:42+00:00` to `2009-06-27 11:12:37+00:00` only contains staypoint records (from staypoint id 396, 397, 398, all non-activity).

the code to reproduce this error using `geolife_1.csv` attached (unzip!).
[geolife_1.zip](https://github.com/mie-lab/trackintel/files/6796006/geolife_1.zip)


```python
pfs = ti.read_positionfixes_csv("geolife_1.csv", index_col="id")

pfs, stps = pfs.as_positionfixes.generate_staypoints(
    method="sliding", dist_threshold=25, time_threshold=5, print_progress=True
)
stps = stps.as_staypoints.create_activity_flag(time_threshold=15)
pfs, tpls = pfs.as_positionfixes.generate_triplegs(stps)

stps.to_csv("stps.csv")
tpls.to_csv("tpls.csv")

# generate trips and a joint staypoint/triplegs dataframe
stps, tpls, trips = ti.preprocessing.generate_trips(stps, tpls, gap_threshold=15)
stps_, tpls_, trips_ = _generate_trips_old(stps, tpls, gap_threshold=15, print_progress=True)

trips.to_csv("trips_new.csv")
trips_.to_csv("trips_old.csv")
```

The error record can be found in `trips_new.csv`, and comparing the start and end time with `stps.csv` and `tpls.csv`

**Comment by bifbof** [comment]
Thanks for running it. I think I may know the reason for that to quote myself 1 month ago:
> There is one thing I am not sure about, if we filter out trips without triplegs as we say in the docstring.

Changing this behavior should be pretty easy, I'll take a look at it.

**Review by hong2223 — COMMENTED** [review]
Sorry for taking a bit longer than expected. I observed another error when running this on a bigger subset of Geolife. The generated trips have two records different from the original function
Original (old function):
| id | user_id | started_at | finished_at | origin_staypoint_id | destination_staypoint_id |
|---|---|---|---|---|---|
|698 |52 |2008-04-29 08:21:34+00:00 | 2008-05-20 10:53:46+00:00 || 646.0
|5472 | 88  | 2008-04-11 08:38:23+00:00 | 2008-09-26 11:13:15+00:00|||

Now (new):
| id | user_id | started_at | finished_at | origin_staypoint_id | destination_staypoint_id |
|---|---|---|---|---|---|
|689 |52 |2008-04-29 08:21:34+00:00 | 2008-10-30 07:37:25+00:00 | | 646.0
|5472 | 88  | 2008-04-11 08:38:23+00:00 | 2008-10-12 08:11:36+00:00 | | |

Note that the `finished_at` time is actually the last tripleg `finished_at` for that user. These records have super long time spans and I am not sure what causes this discrepancy.


```python
pfs, mode_labels = read_geolife("./geolife/Data", print_progress=True)
# get user with mode labels
user_with_mode = []
for key, value in mode_labels.items():
    if len(value) > 0:
        user_with_mode.append(key)
# only select user who has mode labels
pfs = pfs.loc[pfs["user_id"].isin(user_with_mode)].copy()

# generate staypoints, triplegs and trips
pfs, spts = pfs.as_positionfixes.generate_staypoints(time_threshold=5.0, print_progress=True)
pfs, tpls = pfs.as_positionfixes.generate_triplegs(spts)
tpls = geolife_add_modes_to_triplegs(tpls, mode_labels)

spts = ti.analysis.labelling.create_activity_flag(spts, time_threshold=15)

stps, tpls, trips = ti.preprocessing.triplegs.generate_trips(spts, tpls, gap_threshold=15)
stps_, tpls_, trips_ = generate_trips_new(spts, tpls, gap_threshold=15)

# print(trips, trips_)
trips.to_csv("trips_old.csv")
trips_.to_csv("trips_new.csv")
spts.to_csv("spts.csv")
tpls.to_csv("tpls.csv")
assert_frame_equal(trips, trips_, check_like=True, check_index_type=False)
assert_frame_equal(stps, stps_, check_like=True, check_index_type=False)
assert_frame_equal(tpls, tpls_, check_like=True, check_index_type=False)
```

**Comment by henrymartin1** [comment]
> 
> 
> Sorry for taking a bit longer than expected. I observed another error when running this on a bigger subset of Geolife. The generated trips have two records different from the original function
> Original (old function):
> id 	user_id 	started_at 	finished_at 	origin_staypoint_id 	destination_staypoint_id
> 698 	52 	2008-04-29 08:21:34+00:00 	2008-05-20 10:53:46+00:00 		646.0
> 5472 	88 	2008-04-11 08:38:23+00:00 	2008-09-26 11:13:15+00:00 		
> 
> Now (new):
> id 	user_id 	started_at 	finished_at 	origin_staypoint_id 	destination_staypoint_id
> 689 	52 	2008-04-29 08:21:34+00:00 	2008-10-30 07:37:25+00:00 		646.0
> 5472 	88 	2008-04-11 08:38:23+00:00 	2008-10-12 08:11:36+00:00 		
> 
> Note that the `finished_at` time is actually the last tripleg `finished_at` for that user. These records have super long time spans and I am not sure what causes this discrepancy.
> 
> ```python
> pfs, mode_labels = read_geolife("./geolife/Data", print_progress=True)
> # get user with mode labels
> user_with_mode = []
> for key, value in mode_labels.items():
>     if len(value) > 0:
>         user_with_mode.append(key)
> # only select user who has mode labels
> pfs = pfs.loc[pfs["user_id"].isin(user_with_mode)].copy()
> 
> # generate staypoints, triplegs and trips
> pfs, spts = pfs.as_positionfixes.generate_staypoints(time_threshold=5.0, print_progress=True)
> pfs, tpls = pfs.as_positionfixes.generate_triplegs(spts)
> tpls = geolife_add_modes_to_triplegs(tpls, mode_labels)
> 
> spts = ti.analysis.labelling.create_activity_flag(spts, time_threshold=15)
> 
> stps, tpls, trips = ti.preprocessing.triplegs.generate_trips(spts, tpls, gap_threshold=15)
> stps_, tpls_, trips_ = generate_trips_new(spts, tpls, gap_threshold=15)
> 
> # print(trips, trips_)
> trips.to_csv("trips_old.csv")
> trips_.to_csv("trips_new.csv")
> spts.to_csv("spts.csv")
> tpls.to_csv("tpls.csv")
> assert_frame_equal(trips, trips_, check_like=True, check_index_type=False)
> assert_frame_equal(stps, stps_, check_like=True, check_index_type=False)
> assert_frame_equal(tpls, tpls_, check_like=True, check_index_type=False)
> ```

ah wow, that actually looks really wrong for both cases... Thanks for the example, I'll have a look

**Comment by henrymartin1** [comment]
Note for future bug reports
If possible, include the import statements in code-snippets :-)

**Comment by bifbof** [comment]
Thanks Ye for finding that error. But I have problems reproducing it.
I work only a subset of the trackintel data of users 50-55 and 80-90 more my laptop cannot handle. ^^
And run on that subset I don't get that error in the trips.
But I found instead found a bug in `geolife_add_modes_to_triplegs` (#273) but that one _should_ have no influence over the outcome.

**Comment by hong2223** [comment]
**TL;DR**: the current version is good, we can merge this now.

I doubled checked the result of generate_trips() after the #283 fix. Now, I cannot reproduce the original bug as reported, guess this is a really special case and we can leave it for now.

I do find another discrepancy when comparing this to the old function, but I think it's a bug in the old function. The old function sometimes (10 times in Geolife) assigns an `origin_staypoint_id` to trips when the previous staypoint and the first tripleg has a temporal gap larger than the defined `gap_threshold` (correct assignment should be pd.NA). This is because in Line [237](https://github.com/mie-lab/trackintel/blob/47c4d236cd18ba204db9c1772e9091230c730f5a/trackintel/preprocessing/triplegs.py#L237) we directly assign the current row to origin_activity, without judging whether the origin_activity and first tripleg has a gap.

The bug can already be observed in the generated trip_id 52 of user 0

```python
import trackintel as ti
from pandas.testing import assert_frame_equal

# data from user 0
pfs, mode_labels = read_geolife("./Data", print_progress=True)

pfs, stps = pfs.as_positionfixes.generate_staypoints(time_threshold=5, print_progress=True)
pfs, tpls = pfs.as_positionfixes.generate_triplegs(stps)
stps = stps.as_staypoints.create_activity_flag(time_threshold=15)

stps_1, tpls_1, trips_1 = generate_trips_new(stps, tpls, gap_threshold=15)
stps_2, tpls_2, trips_2 = _generate_trips_old(stps, tpls, gap_threshold=15, print_progress=True)

# False when considering origin_staypoint_id
assert_frame_equal(trips_1, trips_2, check_like=True, check_index_type=False)
# True when only considering time
assert_frame_equal(trips_1.iloc[:,:3], trips_2.iloc[:,:3], check_like=True, check_index_type=False)
```

**Review by hong2223 — APPROVED** [review]

**Comment by henrymartin1** [comment]
Great work everybody! Thanks @bifbof for most of the implementation and special thanks to @hong2223 for the thorough review!
